### PR TITLE
feat(i18n): manage security domain dictionaries in the gateway

### PR DESCRIPTION
**Pull request by tcompiegne**
closes gravitee-io/issues#8040

## :memo: Test scenarios 

1. Create a new en dictionary
2. For this dictionary, create one translation  `login.page.title`  : 'New Login title`
3. Try to login to your application, the Login title page should be : **New Login title**

### CODE REVIEW

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/manager/dictionary/I18nDictionaryManager.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+ */
+package io.gravitee.am.gateway.handler.manager.dictionary;
+
+import io.gravitee.common.service.Service;
+
+/**
+ * @author Titouan COMPIEGNE (titouan.compiegne at graviteesource.com)
+ * @author GraviteeSource Team
+ */
+public interface I18nDictionaryManager extends Service {
```
Comment (farmborough) [inline gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/manager/dictionary/I18nDictionaryManager.java:24]: I don't understand why this interface is needed, would have been better for `I18nDictionaryManager` to be a class that implements the `Service` marker interface IMO.

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/manager/dictionary/impl/I18nDictionaryManagerImpl.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+    @Autowired
+    private GraviteeMessageResolver graviteeMessageResolver;
+
+    @Override
+    public void afterPropertiesSet() throws Exception {
+        logger.info("Initializing i18n dictionaries for domain {}", domain.getName());
+        i18nDictionaryRepository.findAll(ReferenceType.DOMAIN, domain.getId())
+                .subscribe(
+                        i18nDictionary -> System.out.println(i18nDictionary),
```
Comment (farmborough) [inline gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/manager/dictionary/impl/I18nDictionaryManagerImpl.java:64]: Can be replaced with method reference. The error is being logged anyway so I don't think we need `System.out` anyway.

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/vertx/view/thymeleaf/DomainBasedDictionaryProvider.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+import java.util.Properties;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static java.util.Optional.ofNullable;
+
+/**
+ * @author Titouan COMPIEGNE (titouan.compiegne at graviteesource.com)
+ * @author GraviteeSource Team
+ */
+public class DomainBasedDictionaryProvider implements DictionaryProvider {
```
Comment (farmborough) [inline gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/vertx/view/thymeleaf/DomainBasedDictionaryProvider.java:32]: Unit test?

**Review by farmborough — CHANGES_REQUESTED** [review]

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/vertx/view/thymeleaf/DomainBasedDictionaryProvider.java (hunk 3 above)
Comment (tcompiegne) [reply gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/vertx/view/thymeleaf/DomainBasedDictionaryProvider.java:32]: FP

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/manager/dictionary/impl/I18nDictionaryManagerImpl.java (hunk 2 above)
Comment (tcompiegne) [reply gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/manager/dictionary/impl/I18nDictionaryManagerImpl.java]: FP

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/manager/dictionary/impl/I18nDictionaryManagerImpl.java [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * @author Titouan COMPIEGNE (titouan.compiegne at graviteesource.com)
+ * @author GraviteeSource Team
+ */
+public class I18nDictionaryManagerImpl extends AbstractService implements I18nDictionaryManager, InitializingBean, EventListener<I18nDictionaryEvent, Payload> {
```
[the remark this answers is not in the document]
Comment (tcompiegne) [reply gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/manager/dictionary/impl/I18nDictionaryManagerImpl.java]: FP

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/test/java/io/gravitee/am/gateway/handler/manager/dictionary/I18nDictionaryManagerTest.java [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+        verify(graviteeMessageResolver, never()).updateDictionary(any());
+    }
+
+    @Test
+    public void shouldDeploy() {
+        shouldDeploy(I18nDictionaryEvent.DEPLOY);
+    }
+
+    @Test
+    public void shouldUpdate() {
```
Comment (ashraf706) [inline gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/test/java/io/gravitee/am/gateway/handler/manager/dictionary/I18nDictionaryManagerTest.java:91]: Informative: Test without assertion violates SonarQube S2699 (Tests should include assertions) rule (we may need to suppress the rule in that case). Possibly we could use assertDoesNotThrow if it were Junit5 [never tried though :) ].

**Review by ashraf706 — APPROVED** [review]

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/manager/dictionary/I18nDictionaryManager.java (hunk 1 above)
Comment (tcompiegne) [reply gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/manager/dictionary/I18nDictionaryManager.java]: FP

**Pull request merged**